File: backend/app/api/writer_agent.py
"""Writer Agent API blueprint."""
import json
import traceback

from flask import Blueprint, Response, jsonify, request

writer_agent_bp = Blueprint("writer_agent", __name__)


def _error_response(exc, status_code=500):
    return jsonify({
        "success": False,
        "error": str(exc),
        "traceback": traceback.format_exc(),
    }), status_code


# ---- Core Writing (SSE) ----

@writer_agent_bp.route("/run", methods=["POST"])
def run_writer_agent():
    """Execute a writing task. Returns SSE stream."""
    data = request.get_json() or {}
    from ..services.writer_agent.orchestrator import WriterOrchestrator
    orchestrator = WriterOrchestrator()

    def event_stream():
        try:
            for event in orchestrator.run(data):
                yield f"data: {json.dumps(event, ensure_ascii=False)}\n\n"
        except Exception as exc:
            yield f"data: {json.dumps({'type': 'error', 'message': str(exc)}, ensure_ascii=False)}\n\n"

    return Response(
        event_stream(),
        mimetype="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no", "Connection": "keep-alive"},
    )


# ---- Scene CRUD ----

@writer_agent_bp.route("/scenes/<chapter_id>", methods=["GET"])
def list_scenes(chapter_id):
    try:
        project_id = request.args.get("project_id", "")
        from ..services.writer_agent.scene_service import SceneService
        scenes = SceneService().list_scenes(project_id, chapter_id)
        return jsonify({"success": True, "data": scenes})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/scenes/detail/<scene_id>", methods=["GET"])
def get_scene_detail(scene_id):
    try:
        project_id = request.args.get("project_id", "")
        from ..services.writer_agent.scene_service import SceneService
        scene = SceneService().get_scene(project_id, scene_id)
        return jsonify({"success": True, "data": scene})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/scenes/<scene_id>", methods=["PUT"])
def update_scene(scene_id):
    try:
        data = request.get_json() or {}
        project_id = data.get("project_id", "")
        from ..services.writer_agent.scene_service import SceneService
        result = SceneService().update_scene(
            project_id, scene_id,
            content=data.get("content"),
            title=data.get("title"),
        )
        return jsonify({"success": True, "data": result})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/scenes/<scene_id>", methods=["DELETE"])
def delete_scene(scene_id):
    try:
        project_id = request.args.get("project_id", "")
        from ..services.writer_agent.scene_service import SceneService
        SceneService().delete_scene(project_id, scene_id)
        return jsonify({"success": True})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/scenes/<chapter_id>/reorder", methods=["POST"])
def reorder_scenes(chapter_id):
    try:
        data = request.get_json() or {}
        project_id = data.get("project_id", "")
        scene_ids = data.get("scene_ids", [])
        from ..services.writer_agent.scene_service import SceneService
        SceneService().reorder_scenes(project_id, chapter_id, scene_ids)
        return jsonify({"success": True})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/scenes/<chapter_id>/compile", methods=["POST"])
def compile_chapter_from_scenes(chapter_id):
    try:
        data = request.get_json() or {}
        project_id = data.get("project_id", "")
        from ..services.writer_agent.scene_service import SceneService
        result = SceneService().compile_chapter(project_id, chapter_id)
        return jsonify({"success": True, "data": result})
    except Exception as exc:
        return _error_response(exc)


# ---- Preset CRUD ----

@writer_agent_bp.route("/presets", methods=["GET"])
def list_presets():
    try:
        project_id = request.args.get("project_id", "")
        from ..services.writer_agent.preset_service import PresetService
        presets = PresetService().list_presets(project_id)
        return jsonify({"success": True, "data": presets})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/presets", methods=["POST"])
def create_preset():
    try:
        data = request.get_json() or {}
        from ..services.writer_agent.preset_service import PresetService
        result = PresetService().create_preset(
            project_id=data.get("project_id"),
            name=data["name"],
            system_prompt=data["system_prompt"],
            description=data.get("description", ""),
            is_default=data.get("is_default", 0),
        )
        return jsonify({"success": True, "data": result})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/presets/<preset_id>", methods=["PUT"])
def update_preset(preset_id):
    try:
        data = request.get_json() or {}
        project_id = data.get("project_id", "")
        from ..services.writer_agent.preset_service import PresetService
        result = PresetService().update_preset(project_id, preset_id, **{
            k: v for k, v in data.items() if k in ("name", "system_prompt", "description", "is_default")
        })
        return jsonify({"success": True, "data": result})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/presets/<preset_id>", methods=["DELETE"])
def delete_preset(preset_id):
    try:
        project_id = request.args.get("project_id", "")
        from ..services.writer_agent.preset_service import PresetService
        PresetService().delete_preset(project_id, preset_id)
        return jsonify({"success": True})
    except Exception as exc:
        return _error_response(exc)


# ---- Chapter CRUD ----

@writer_agent_bp.route("/chapters/<project_id>", methods=["GET"])
def list_chapters(project_id):
    try:
        from ..services.writer_agent.chapter_service import ChapterService
        chapters = ChapterService().list_chapters(project_id)
        return jsonify({"success": True, "data": chapters})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/chapters/<project_id>", methods=["POST"])
def create_chapter(project_id):
    try:
        data = request.get_json() or {}
        from ..services.writer_agent.chapter_service import ChapterService
        result = ChapterService().create_chapter(
            project_id,
            title=data.get("title", ""),
            chapter_order=data.get("chapter_order"),
        )
        return jsonify({"success": True, "data": result})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/chapters/detail/<chapter_id>", methods=["PUT"])
def update_chapter(chapter_id):
    try:
        data = request.get_json() or {}
        project_id = data.get("project_id", "")
        from ..services.writer_agent.chapter_service import ChapterService
        result = ChapterService().update_chapter(project_id, chapter_id, **{
            k: v for k, v in data.items()
            if k in ("title", "summary", "outline_json", "timeline_note", "open_threads_json", "pov_character")
        })
        return jsonify({"success": True, "data": result})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/chapters/detail/<chapter_id>", methods=["DELETE"])
def delete_chapter(chapter_id):
    try:
        project_id = request.args.get("project_id", "")
        from ..services.writer_agent.chapter_service import ChapterService
        ChapterService().delete_chapter(project_id, chapter_id)
        return jsonify({"success": True})
    except Exception as exc:
        return _error_response(exc)


# ---- Manuscript ----

@writer_agent_bp.route("/manuscript/<project_id>/commit", methods=["POST"])
def commit_to_manuscript(project_id):
    try:
        data = request.get_json() or {}
        content = data.get("content", "")
        if not content.strip():
            return jsonify({"success": False, "error": "内容不能为空"}), 400
        from ..services.writer_agent.manuscript_service import ManuscriptService
        result = ManuscriptService().commit(
            project_id,
            content=content,
            source_scene_id=data.get("source_scene_id"),
            insert_after_block_id=data.get("insert_after_block_id"),
            chapter_tag=data.get("chapter_tag"),
        )
        return jsonify({"success": True, "data": result})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/manuscript/<project_id>", methods=["GET"])
def list_manuscript(project_id):
    try:
        include_content = request.args.get("include_content", "true").lower() == "true"
        from ..services.writer_agent.manuscript_service import ManuscriptService
        result = ManuscriptService().list_blocks(project_id, include_content)
        return jsonify({"success": True, "data": result})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/manuscript/block/<block_id>", methods=["PUT"])
def update_manuscript_block(block_id):
    try:
        data = request.get_json() or {}
        project_id = data.get("project_id", "")
        kwargs = {k: v for k, v in data.items() if k in ("content", "chapter_tag")}
        from ..services.writer_agent.manuscript_service import ManuscriptService
        result = ManuscriptService().update_block(project_id, block_id, **kwargs)
        return jsonify({"success": True, "data": result})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/manuscript/block/<block_id>", methods=["DELETE"])
def delete_manuscript_block(block_id):
    try:
        project_id = request.args.get("project_id", "")
        from ..services.writer_agent.manuscript_service import ManuscriptService
        ManuscriptService().delete_block(project_id, block_id)
        return jsonify({"success": True})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/manuscript/<project_id>/reorder", methods=["PUT"])
def reorder_manuscript(project_id):
    try:
        data = request.get_json() or {}
        block_ids = data.get("block_ids", [])
        from ..services.writer_agent.manuscript_service import ManuscriptService
        ManuscriptService().reorder(project_id, block_ids)
        return jsonify({"success": True})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/manuscript/<project_id>/tag", methods=["PUT"])
def tag_manuscript_blocks(project_id):
    try:
        data = request.get_json() or {}
        block_ids = data.get("block_ids", [])
        chapter_tag = data.get("chapter_tag", "")
        from ..services.writer_agent.manuscript_service import ManuscriptService
        count = ManuscriptService().tag_blocks(project_id, block_ids, chapter_tag)
        return jsonify({"success": True, "data": {"updated_count": count}})
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/manuscript/<project_id>/export", methods=["GET"])
def export_manuscript(project_id):
    try:
        fmt = request.args.get("format", "txt")
        from ..services.writer_agent.manuscript_service import ManuscriptService
        text = ManuscriptService().export_text(project_id, fmt)
        filename = f"manuscript.{fmt}"
        return Response(
            text,
            mimetype="text/plain; charset=utf-8",
            headers={"Content-Disposition": f"attachment; filename={filename}"},
        )
    except Exception as exc:
        return _error_response(exc)


@writer_agent_bp.route("/manuscript/<project_id>/continuation-context", methods=["GET"])
def get_continuation_context(project_id):
    try:
        token_budget = int(request.args.get("token_budget", "8000"))
        last_block_id = request.args.get("last_block_id") or None
        from ..services.writer_agent.manuscript_context_builder import build_continuation_context
        result = build_continuation_context(project_id, token_budget, last_block_id=last_block_id)
        return jsonify({"success": True, "data": result})
    except Exception as exc:
        return _error_response(exc)


# ---- Migration ----

@writer_agent_bp.route("/migrate/<project_id>", methods=["POST"])
def migrate_project(project_id):
    try:
        from ..services.writer_agent.novel_db_migration import migrate_project as do_migrate
        result = do_migrate(project_id)
        return jsonify({"success": True, "data": result})
    except Exception as exc:
        return _error_response(exc)

File: frontend/src/api/writer-agent.ts
/**
 * Writer Agent API client — writing workbench, scenes, chapters, manuscripts,
 * presets, outline versioning, and world data updates.
 */

import { del, get, post, put } from "@/api/http";
import type { ApiResponse } from "@/api/http";
import { postSSE } from "@/api/sse";
import type { SSEHandlers } from "@/api/sse";

/* ---------- Core writing (SSE) ---------- */

export function runWriterAgent(
  payload: unknown,
  handlers: SSEHandlers,
  signal?: AbortSignal,
): Promise<void> {
  return postSSE("/api/writer-agent/run", payload, handlers, signal);
}

export function updateWorldData(
  payload: unknown,
  handlers: SSEHandlers,
  signal?: AbortSignal,
): Promise<void> {
  return postSSE("/api/writer-agent/world-update", payload, handlers, signal);
}

/** Apply user-selected reviewer suggestions — SSE stream of the rewrite. */
export function applyReviewer(
  payload: unknown,
  handlers: SSEHandlers,
  signal?: AbortSignal,
): Promise<void> {
  return postSSE("/api/writer-agent/apply-reviewer", payload, handlers, signal);
}

/* ---------- One-click buttons (W-3 §4.4) ---------- */

export function oneClickCompleteOutline(
  payload: { project_id: string; chapter_id: string; chapter_order?: number },
  handlers: SSEHandlers,
  signal?: AbortSignal,
): Promise<void> {
  return postSSE(
    "/api/writer-agent/one-click/complete-outline",
    payload,
    handlers,
    signal,
  );
}

export function oneClickAlignWords(
  payload: {
    project_id: string;
    chapter_id: string;
    chapter_order?: number;
    target_word_count: number;
    tolerance_pct?: number;
  },
  handlers: SSEHandlers,
  signal?: AbortSignal,
): Promise<void> {
  return postSSE(
    "/api/writer-agent/one-click/align-words",
    payload,
    handlers,
    signal,
  );
}

export function oneClickScanLexicon(
  payload: {
    project_id: string;
    chapter_id: string;
    chapter_order?: number;
    lexicon_asset_ids?: string[];
  },
  handlers: SSEHandlers,
  signal?: AbortSignal,
): Promise<void> {
  return postSSE(
    "/api/writer-agent/one-click/scan-lexicon",
    payload,
    handlers,
    signal,
  );
}

export function oneClickFillRelationships(
  payload: { project_id: string; chapter_id: string; chapter_order?: number },
  handlers: SSEHandlers,
  signal?: AbortSignal,
): Promise<void> {
  return postSSE(
    "/api/writer-agent/one-click/fill-relationships",
    payload,
    handlers,
    signal,
  );
}

export function oneClickContinueChapter(
  payload: {
    project_id: string;
    chapter_id: string;
    chapter_order?: number;
    last_block_id?: string;
    target_word_count?: number;
    preset_id?: string;
  },
  handlers: SSEHandlers,
  signal?: AbortSignal,
): Promise<void> {
  return postSSE(
    "/api/writer-agent/one-click/continue-chapter",
    payload,
    handlers,
    signal,
  );
}

/* ---------- Scene CRUD ---------- */

export function getScenes(chapterId: string, projectId: string): Promise<ApiResponse> {
  return get(`/api/writer-agent/scenes/${chapterId}?project_id=${projectId}`);
}

export function createScene(chapterId: string, payload: unknown): Promise<ApiResponse> {
  return post(`/api/writer-agent/scenes/${chapterId}`, payload);
}

export function getSceneDetail(sceneId: string, projectId: string): Promise<ApiResponse> {
  return get(`/api/writer-agent/scenes/detail/${sceneId}?project_id=${projectId}`);
}

export function updateScene(sceneId: string, payload: unknown): Promise<ApiResponse> {
  return put(`/api/writer-agent/scenes/${sceneId}`, payload);
}

export function deleteScene(sceneId: string, projectId: string): Promise<ApiResponse> {
  return del(`/api/writer-agent/scenes/${sceneId}?project_id=${projectId}`);
}

export function reorderScenes(chapterId: string, payload: unknown): Promise<ApiResponse> {
  return post(`/api/writer-agent/scenes/${chapterId}/reorder`, payload);
}

export function compileChapter(chapterId: string, payload: unknown): Promise<ApiResponse> {
  return post(`/api/writer-agent/scenes/${chapterId}/compile`, payload);
}

/* ---------- Preset CRUD ---------- */

export function getPresets(projectId?: string): Promise<ApiResponse> {
  return get(`/api/writer-agent/presets?project_id=${projectId || ""}`);
}

export function createPreset(payload: unknown): Promise<ApiResponse> {
  return post("/api/writer-agent/presets", payload);
}

export function updatePreset(presetId: string, payload: unknown): Promise<ApiResponse> {
  return put(`/api/writer-agent/presets/${presetId}`, payload);
}

export function deletePreset(presetId: string, projectId?: string): Promise<ApiResponse> {
  return del(`/api/writer-agent/presets/${presetId}?project_id=${projectId || ""}`);
}

/* ---------- Chapter CRUD ---------- */

export function getChapters(projectId: string): Promise<ApiResponse> {
  return get(`/api/writer-agent/chapters/${projectId}`);
}

export function createChapter(projectId: string, payload: unknown): Promise<ApiResponse> {
  return post(`/api/writer-agent/chapters/${projectId}`, payload);
}

export function updateChapter(chapterId: string, payload: unknown): Promise<ApiResponse> {
  return put(`/api/writer-agent/chapters/detail/${chapterId}`, payload);
}

export function deleteChapter(chapterId: string, projectId: string): Promise<ApiResponse> {
  return del(`/api/writer-agent/chapters/detail/${chapterId}?project_id=${projectId}`);
}

/* ---------- Manuscript ---------- */

export function commitToManuscript(
  projectId: string,
  payload: unknown,
): Promise<ApiResponse> {
  return post(`/api/writer-agent/manuscript/${projectId}/commit`, payload);
}

export function getManuscript(
  projectId: string,
  includeContent = true,
): Promise<ApiResponse> {
  return get(
    `/api/writer-agent/manuscript/${projectId}?include_content=${includeContent}`,
  );
}

export function updateManuscriptBlock(
  blockId: string,
  payload: unknown,
): Promise<ApiResponse> {
  return put(`/api/writer-agent/manuscript/block/${blockId}`, payload);
}

export function deleteManuscriptBlock(
  blockId: string,
  projectId: string,
): Promise<ApiResponse> {
  return del(`/api/writer-agent/manuscript/block/${blockId}?project_id=${projectId}`);
}

export function reorderManuscriptBlocks(
  projectId: string,
  blockIds: string[],
): Promise<ApiResponse> {
  return put(`/api/writer-agent/manuscript/${projectId}/reorder`, {
    block_ids: blockIds,
  });
}

export function tagManuscriptBlocks(
  projectId: string,
  payload: unknown,
): Promise<ApiResponse> {
  return put(`/api/writer-agent/manuscript/${projectId}/tag`, payload);
}

export function moveManuscriptBlock(
  blockId: string,
  payload: unknown,
): Promise<ApiResponse> {
  return put(`/api/writer-agent/manuscript/block/${blockId}/move`, payload);
}

/**
 * Export manuscript as a downloadable Blob.
 * Uses raw fetch() — does NOT go through the http wrapper.
 */
export async function exportManuscript(
  projectId: string,
  format = "txt",
): Promise<Blob> {
  const response = await fetch(
    `/api/writer-agent/manuscript/${projectId}/export?format=${format}`,
  );
  if (!response.ok) throw new Error(`导出失败: ${response.status}`);
  return response.blob();
}

/* ---------- Continuation context ---------- */

export interface ContinuationContextOptions {
  tokenBudget?: number;
  lastBlockId?: string;
}

export function getContinuationContext(
  projectId: string,
  { tokenBudget = 8000, lastBlockId = "" }: ContinuationContextOptions = {},
): Promise<ApiResponse> {
  let url = `/api/writer-agent/manuscript/${projectId}/continuation-context?token_budget=${tokenBudget}`;
  if (lastBlockId) url += `&last_block_id=${encodeURIComponent(lastBlockId)}`;
  return get(url);
}

/* ---------- Outline versions ---------- */

export function getOutlineVersions(
  chapterId: string,
  projectId: string,
): Promise<ApiResponse> {
  return get(
    `/api/writer-agent/chapters/detail/${chapterId}/outline-versions?project_id=${projectId}`,
  );
}

export function getOutlineVersion(
  chapterId: string,
  versionId: string,
  projectId: string,
): Promise<ApiResponse> {
  return get(
    `/api/writer-agent/chapters/detail/${chapterId}/outline-versions/${versionId}?project_id=${projectId}`,
  );
}

export function restoreOutlineVersion(
  chapterId: string,
  versionId: string,
  projectId: string,
): Promise<ApiResponse> {
  return post(
    `/api/writer-agent/chapters/detail/${chapterId}/outline-versions/${versionId}/restore`,
    { project_id: projectId },
  );
}

/* ---------- Migration ---------- */

export function migrateProject(projectId: string): Promise<ApiResponse> {
  return post(`/api/writer-agent/migrate/${projectId}`);
}

/* ---------- Book-run (multi-chapter agent task) ---------- */

export interface BookPlanInput {
  project_id: string;
  title: string;
  chapter_count: number;
  per_chapter_word_target: number;
  word_tolerance_pct?: number;
  overall_direction?: string;
  global_brief?: string;
  start_chapter_order?: number;
  forbidden_lexicon_asset_ids?: string[];
  style_asset_ids?: string[];
  preset_id?: string | null;
}

export function listBookPlans(projectId: string): Promise<ApiResponse> {
  return get(`/api/writer-agent/book-plans?project_id=${encodeURIComponent(projectId)}`);
}

export function createBookPlan(payload: BookPlanInput): Promise<ApiResponse> {
  return post("/api/writer-agent/book-plans", payload);
}

export function getBookPlan(planId: string): Promise<ApiResponse> {
  return get(`/api/writer-agent/book-plans/${planId}`);
}

export function updateBookPlan(planId: string, payload: Partial<BookPlanInput>): Promise<ApiResponse> {
  return put(`/api/writer-agent/book-plans/${planId}`, payload);
}

export function deleteBookPlan(planId: string): Promise<ApiResponse> {
  return del(`/api/writer-agent/book-plans/${planId}`);
}

export function getBookPlanStatus(planId: string): Promise<ApiResponse> {
  return get(`/api/writer-agent/book-plans/${planId}/status`);
}

export function getActiveBookPlan(projectId: string): Promise<ApiResponse> {
  return get(`/api/writer-agent/book-plans/active/by-project?project_id=${encodeURIComponent(projectId)}`);
}

export function runBookRun(
  planId: string,
  handlers: SSEHandlers,
  signal?: AbortSignal,
): Promise<void> {
  return postSSE("/api/writer-agent/book-run", { plan_id: planId }, handlers, signal);
}

export function abortBookRun(planId: string): Promise<ApiResponse> {
  return post("/api/writer-agent/book-run/abort", { plan_id: planId });
}

/* ---------- Forbidden lexicon ---------- */

export interface ForbiddenLexiconEntry {
  pattern: string;
  match_type?: "literal" | "regex" | "phrase";
  category?: "word" | "sentence" | "stylistic";
  severity?: "block" | "warn";
  note?: string;
  whitelist_contexts?: string[];
}

export function listForbiddenLexicons(
  projectId: string,
  includeEntries = false,
): Promise<ApiResponse> {
  return get(
    `/api/writer-agent/forbidden-lexicons?project_id=${encodeURIComponent(projectId)}&include_entries=${includeEntries}`,
  );
}

export function upsertForbiddenLexicon(payload: {
  project_id: string;
  asset_id?: string | null;
  title?: string;
  entries: ForbiddenLexiconEntry[] | string[];
}): Promise<ApiResponse> {
  return post("/api/writer-agent/forbidden-lexicons", payload);
}

export function deleteForbiddenLexicon(
  assetId: string,
  projectId: string,
): Promise<ApiResponse> {
  return del(
    `/api/writer-agent/forbidden-lexicons/${assetId}?project_id=${encodeURIComponent(projectId)}`,
  );
}
